distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+        ActiveMemoryManager chose to disregard the request.
+
+        The current pending (accepted) commands can be inspected on
+        ``self.manager.pending``; this includes the commands previously yielded by this
+        same method.
+
+        The current memory usage on each worker, *downstream of all pending commands*,
+        can be inspected on ``self.manager.workers_memory``.
+        """
+        raise NotImplementedError("Virtual method")
+
+
+class ReplicateReduce(ActiveMemoryManagerPlugin):
+    """Make sure that we don't have more replicas than desired for every key; drop
+    excess replicas
+    """
+
+    def run(self):
+        # TODO this is O(n) to the total number of in-memory tasks on the cluster; it
+        #      could be made faster by automatically attaching it to a key when it goes
+        #      above one replica and detaching it when it goes below.
+        for ts in self.manager.scheduler.tasks.values():
+            desired_replicas = 1  # TODO have a marker on TaskState
+            ndrop = len(ts.who_has) - max(desired_replicas, len(ts.waiters))
+            if ts in self.manager.pending:
+                pending_repl, pending_drop = self.manager.pending[ts]
+                ndrop += len(pending_repl) - len(pending_drop)
+            for _ in range(ndrop):
+                yield "drop", ts, None
+
+
+class ReplicateIncrease(ActiveMemoryManagerPlugin):
+    """Make sure that the number of replicas for a key is, at all times, at least what
+    the user asked for through ``Client.replicate()`` or
+    ``Client.scatter(..., broadcast=True)``. Only the keys that have been listed in one
+    of the above two methods have this plugin attached to them. If a user later invokes
+    ``Client.replicate(..., 1)``, this plugin will be automatically detached at the next
+    AMM iteration.
+    """
+
+    ts: TaskState
+
+    def __init__(self, ts: TaskState):
+        self.ts = ts
+
+    def __repr__(self):
+        return f"ReplicateIncrease({self.ts.key})"
+
+    def run(self):
+        desired_replicas = 1  # TODO have a marker on TaskState
+        if desired_replicas == 1:
+            self.manager.plugins.remove(self)
+        elif self.ts.who_has:
+            for _ in range(desired_replicas - len(self.ts.who_has)):
+                yield "replicate", self.ts, None
+
+
+class WorkerShutdown(ActiveMemoryManagerPlugin):
+    """Replicate somewhere else all unique keys on a worker, preparing for its shutdown.
+    When there are no longer unique keys, this plugin removes itself from the AMM.
+    """
+
+    ws: WorkerState
+
+    def __init__(self, ws: WorkerState):
+        self.ws = ws
+
+    def __repr__(self) -> str:
+        return f"WorkerShutdown({self.ws.address})"
+
+    def run(self):
+        unique = {ts for ts in self.ws.has_what if len(ts.who_has) == 1}
+        for ts in unique:
+            yield "replicate", ts, None
+        if not unique:
+            self.manager.plugins.remove(self)
```

Comment:
missing all the logic that will prevent more tasks from landing on a worker in "shutting down" stage